### ISSUE TITLE
Documentation missing for Dependency Injection in 'new-able' Classes

### ISSUE THREAD

**Issue by gklandes** [angular/angular#24964]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[ ] Feature request
[x] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior

I want to isolate http interactions by creating data access objects from a class so that in a component I might simply get data like this:
```javascript
// dashboard.component
import { AppUser } from './appuser.service'

export class DashboardComponent implements OnInit {
  user: AppUser = new AppUser();

  constructor() { }

  ngOnInit() {
    let id = JSON.parse(window.localStorage.getItem('session')).userId;
    this.user.find(id) // 'find' is from base class
      .subscribe(
        // handle user data
      );
  }
}
```
I have defined a base class and a sub class like this:
```javascript
// base-resource.service
import { HttpClient } from '@angular/common/http';
...
export class BaseResource {
  private fullpath: string;
  protected http: HttpClient;

  constructor (path: string) { //<-- Can't inject here b/c its used in instantiation
    this.fullpath = path;
  }

  find (id): Observable<Object> {
    return this.http.get(this.fullpath + '/' + id); // this line throws Error!
  }
}

// app-user.service
...
export class AppUser extends BaseResource {
  constructor(data?) {
    super('api/appusers');
  }
}
```
This obviously generates an error b/c HttpClient is not actually initialized in BaseResource. I get `ERROR TypeError: Cannot read property 'get' of undefined` from within the base class function.

## Expected behavior

I have reviewed the following documentation
* https://angular.io/guide/dependency-injection 
* https://angular.io/api/core/Injectable

I would expect to find this use case handled in one of the following ways
* explain how to properly accomplish the above scenario (DI in class)
* explain why this scenario is an anti-pattern; including a suggestion for a better method

## Environment

<pre><code>
Angular version: 6.0.4

**Comment by splincode** [comment]
So will not work, you need

```
// base-resource.service
import { HttpClient } from '@angular/common/http';
...
export class BaseResource {
  private fullpath: string;
 
  constructor (private http: HttpClient, path: string) {
    this.fullpath = path;
  }

  find (id): Observable<Object> {
    return this.http.get(this.fullpath + '/' + id); // this line throws Error!
  }
}

// app-user.service
import { HttpClient } from '@angular/common/http';
...
@Injectable() // <--- You need usage current decorator for DI and usage HttpClientModule
export class AppUser extends BaseResource {
  constructor(private http: HttpClient, data) {
    super(http, 'api/appusers');
  }
}


import { AppUser } from './appuser.service'

export class DashboardComponent implements OnInit {
  constructor(private user: AppUser) { }

  ngOnInit() {
    let id = JSON.parse(window.localStorage.getItem('session')).userId;
    this.user.find(id) // 'find' is from base class
      .subscribe(
        // handle user data
      );
  }
}
```

**Comment by jenniferfell** [comment]
@jbogarthyde Something to consider in the context of your current DI work?

**Comment by gklandes** [comment]
@splincode thanks for the suggestions. You suggested the "@Injectable" in the app-user.service, but I actually did have that in the original but omitted it in my initial comment (under the ellipsis) to be brief. 

The injectable declaration does not address the underlying conflict b/t Angular's DI system and the concept of extending classes in [Typescript](http://www.typescriptlang.org/docs/handbook/classes.html). While the 'extends' keyword works as expected, Angular's usage of the constructor's arguments for DI precludes passing in instance data if you attempt to create new instances of a class. This is NOT a negative critique of Angular or its choices in DI implementation, But this seems like something that should be addressed clearly in the docs. While looking for a solution I ran across a large number of SO questions that hinged on this point.

This question is a bit old but is a good example of the misconception with a pretty decent answer giving direction to what might be called the "Angular way":
[Angular2: How to use multiple instances of same Service?](https://stackoverflow.com/questions/38482357/angular2-how-to-use-multiple-instances-of-same-service?noredirect=1&lq=1)

I think a discussion of "Factory Services" under https://angular.io/guide/architecture-services might be a good solution.
